packages/go_router/lib/src/route.dart

```diff
@@ -433,22 +462,339 @@ class ShellRoute extends RouteBase {
+    super.pageBuilder,
+  })  : assert(branches.isNotEmpty),
+        assert(_debugUniqueNavigatorKeys(branches).length == branches.length,
+            'Navigator keys must be unique'),
+        super._(routes: _routes(branches)) {
+    for (int i = 0; i < routes.length; ++i) {
+      final RouteBase route = routes[i];
+      if (route is GoRoute) {
+        assert(route.parentNavigatorKey == null ||
+            route.parentNavigatorKey == branches[i].navigatorKey);
+      }
+    }
+  }
 
-  /// The page builder for a shell route.
+  /// Representations of the different stateful route branches that this
+  /// shell route will manage.
   ///
-  /// Similar to GoRoute pageBuilder, but with an additional child parameter.
-  /// This child parameter is the Widget built by calling the matching
-  /// sub-route's builder.
-  final ShellRoutePageBuilder? pageBuilder;
+  /// Each branch uses a separate [Navigator], identified
+  /// [StatefulShellBranch.navigatorKey].
+  final List<StatefulShellBranch> branches;
 
-  /// The [GlobalKey] to be used by the [Navigator] built for this route.
-  /// All ShellRoutes build a Navigator by default. Child GoRoutes
-  /// are placed onto this Navigator instead of the root Navigator.
+  @override
+  GlobalKey<NavigatorState> navigatorKeyForSubRoute(RouteBase subRoute) {
+    return branchForSubRoute(subRoute).navigatorKey;
+  }
+
+  /// Returns the StatefulShellBranch that is to be used for the specified
+  /// immediate sub-route of this shell route.
+  StatefulShellBranch branchForSubRoute(RouteBase subRoute) {
+    final StatefulShellBranch? branch = branches.firstWhereOrNull(
+        (StatefulShellBranch e) => e.routes.contains(subRoute));
+    assert(branch != null);
+    return branch!;
+  }
+
+  static List<RouteBase> _routes(List<StatefulShellBranch> branches) =>
+      branches.expand((StatefulShellBranch e) => e.routes).toList();
+
+  static Set<GlobalKey<NavigatorState>> _debugUniqueNavigatorKeys(
+          List<StatefulShellBranch> branches) =>
+      Set<GlobalKey<NavigatorState>>.from(
+          branches.map((StatefulShellBranch e) => e.navigatorKey));
+}
+
+/// Representation of a separate branch in a stateful navigation tree, used to
+/// configure [StatefulShellRoute].
+///
+/// The only required argument when creating a StatefulShellBranch is the
+/// sub-routes ([routes]), however sometimes it may be convenient to also
+/// provide a [defaultLocation]. The value of this parameter is used when
+/// loading the branch for the first time (for instance when switching branch
+/// using the goBranch method in [StatefulShellBranchState]). A [navigatorKey]
+/// can be useful to provide in case it's necessary to access the [Navigator]
+/// created for this branch elsewhere.
+@immutable
+class StatefulShellBranch {
+  /// Constructs a [StatefulShellBranch].
+  StatefulShellBranch({
+    required this.routes,
+    GlobalKey<NavigatorState>? navigatorKey,
+    this.defaultLocation,
+    this.name,
```

Comment:
It looks like name, index, navigatorKey are all can be used to look up branchs, Ideally we should only use one to prevent duplicated API.

Reply:
Yes, perhaps all this is a bit redundant, since you anyway could figure out the index of the branch you looking for by other means.